File: adapters/python/usage_example.py
"""
Example usage of Python adapters with the new type-safe architecture
"""

from dynamic_filter_core import FilterDefinition, FilterRequest, ContextAdapter, ConditionAdapterBuilder, Condition
from dynamic_filter_core.validation import Operator
from dynamic_filter_core.test_types import TestPropertyRef, TestPropertyRefImpl

# Django imports
from .django.django_context_adapter import DjangoContextAdapter
from .django.django_condition_adapter_builder import DjangoConditionAdapterBuilder
from .django.django_condition_adapter import DjangoConditionAdapter

# SQLAlchemy imports
from .sqlalchemy.sqlalchemy_context_adapter import SQLAlchemyContextAdapter
from .sqlalchemy.sqlalchemy_condition_adapter_builder import SQLAlchemyConditionAdapterBuilder
from .sqlalchemy.sqlalchemy_condition_adapter import SQLAlchemyConditionAdapter

# Example entity type
from typing import Any

class User:
    """Example entity type."""
    def __init__(self, id: int, name: str, age: int, email: str, status: str):
        self.id = id
        self.name = name
        self.age = age
        self.email = email
        self.status = status

# Example Django condition adapter builder implementation
class ExampleDjangoConditionAdapterBuilder(DjangoConditionAdapterBuilder[User, TestPropertyRef]):
    def build(self, ref: TestPropertyRef, op: Operator, value: Any) -> DjangoConditionAdapter[User]:
        from django.db.models import Q
        
        # Convert enum key to entity field
        field_name = ref.name.lower().replace('_', '_')
        
        if op == Operator.EQUALS:
            q_object = Q(**{field_name: value})
        elif op == Operator.LIKE:
            q_object = Q(**{f"{field_name}__icontains": value})
        elif op == Operator.GREATER_THAN:
            q_object = Q(**{f"{field_name}__gt": value})
        else:
            raise ValueError(f"Unsupported operator: {op}")
        
        return DjangoConditionAdapter[User](q_object)

# Example SQLAlchemy condition adapter builder implementation
class ExampleSQLAlchemyConditionAdapterBuilder(SQLAlchemyConditionAdapterBuilder[User, TestPropertyRef]):
    def build(self, ref: TestPropertyRef, op: Operator, value: Any) -> SQLAlchemyConditionAdapter[User]:
        from sqlalchemy import and_, or_, not_
        
        # Convert enum key to entity field
        field_name = ref.name.lower().replace('_', '_')
        
        if op == Operator.EQUALS:
            condition = getattr(User, field_name) == value
        elif op == Operator.LIKE:
            condition = getattr(User, field_name).like(f"%{value}%")
        elif op == Operator.GREATER_THAN:
            condition = getattr(User, field_name) > value
        else:
            raise ValueError(f"Unsupported operator: {op}")
        
        return SQLAlchemyConditionAdapter[User](condition)

# Example usage
def demonstrate_adapters():
    # Create filter definitions
    name_filter = FilterDefinition[TestPropertyRef](
        ref=TestPropertyRef.USER_NAME,
        operator=Operator.LIKE,
        value="John%"
    )

    age_filter = FilterDefinition[TestPropertyRef](
        ref=TestPropertyRef.USER_AGE,
        operator=Operator.GREATER_THAN,
        value=18
    )

    # Django adapter usage
    django_builder = ExampleDjangoConditionAdapterBuilder()
    django_adapter = DjangoContextAdapter[User, TestPropertyRef](
        django_builder, 
        TestPropertyRefImpl
    )

    django_adapter.add_condition("nameFilter", name_filter)
    django_adapter.add_condition("ageFilter", age_filter)

    # SQLAlchemy adapter usage
    sqlalchemy_builder = ExampleSQLAlchemyConditionAdapterBuilder()
    sqlalchemy_adapter = SQLAlchemyContextAdapter[User, TestPropertyRef](
        sqlalchemy_builder,
        TestPropertyRefImpl
    )

    sqlalchemy_adapter.add_condition("nameFilter", name_filter)
    sqlalchemy_adapter.add_condition("ageFilter", age_filter)

    print("✅ Python adapters validation successful!")
    
    return { 
        "django_adapter": django_adapter, 
        "sqlalchemy_adapter": sqlalchemy_adapter 
    }

if __name__ == "__main__":
    demonstrate_adapters()

File: adapters/typescript/usage-example.ts
/**
 * Example usage of TypeScript adapters with the new type-safe architecture
 */

import { FilterDefinition, FilterRequest } from '@cyfko/dynamic-filter-core';
import { Operator } from '@cyfko/dynamic-filter-core';
import { TestPropertyRef, TestPropertyRefImpl } from '../../core/typescript/src/test-types';

// TypeORM imports
import { TypeORMContextAdapter } from './typeorm/src/typeorm-context-adapter';
import { TypeORMConditionAdapterBuilder } from './typeorm/src/typeorm-condition-adapter-builder';
import { TypeORMConditionAdapter } from './typeorm/src/typeorm-condition-adapter';

// Prisma imports
import { PrismaContextAdapter } from './prisma/src/prisma-context-adapter';
import { PrismaConditionAdapterBuilder } from './prisma/src/prisma-condition-adapter-builder';
import { PrismaConditionAdapter } from './prisma/src/prisma-condition-adapter';

// Example entity type
interface User {
  id: number;
  name: string;
  age: number;
  email: string;
  status: string;
}

// Example TypeORM condition adapter builder implementation
class ExampleTypeORMConditionAdapterBuilder implements TypeORMConditionAdapterBuilder<User, typeof TestPropertyRefImpl> {
  build(ref: keyof typeof TestPropertyRefImpl, op: Operator, value: any): TypeORMConditionAdapter<User> {
    return new TypeORMConditionAdapter<User>((qb) => {
      const propertyName = String(ref).toLowerCase(); // Convert enum key to entity field
      
      switch (op) {
        case Operator.EQUALS:
          return qb.andWhere(`entity.${propertyName} = :${propertyName}`, { [propertyName]: value });
        case Operator.LIKE:
          return qb.andWhere(`entity.${propertyName} LIKE :${propertyName}`, { [propertyName]: value });
        case Operator.GREATER_THAN:
          return qb.andWhere(`entity.${propertyName} > :${propertyName}`, { [propertyName]: value });
        default:
          throw new Error(`Unsupported operator: ${op}`);
      }
    });
  }
}

// Example Prisma condition adapter builder implementation
class ExamplePrismaConditionAdapterBuilder implements PrismaConditionAdapterBuilder<User, typeof TestPropertyRefImpl> {
  build(ref: keyof typeof TestPropertyRefImpl, op: Operator, value: any): PrismaConditionAdapter<User> {
    const propertyName = String(ref).toLowerCase(); // Convert enum key to entity field
    
    switch (op) {
      case Operator.EQUALS:
        return new PrismaConditionAdapter<User>({ [propertyName]: { equals: value } });
      case Operator.LIKE:
        return new PrismaConditionAdapter<User>({ [propertyName]: { contains: value } });
      case Operator.GREATER_THAN:
        return new PrismaConditionAdapter<User>({ [propertyName]: { gt: value } });
      default:
        throw new Error(`Unsupported operator: ${op}`);
    }
  }
}

// Example usage
function demonstrateAdapters() {
  // Create filter definitions
  const nameFilter: FilterDefinition<typeof TestPropertyRefImpl> = {
    ref: TestPropertyRef.USER_NAME,
    operator: Operator.LIKE,
    value: "John%"
  };

  const ageFilter: FilterDefinition<typeof TestPropertyRefImpl> = {
    ref: TestPropertyRef.USER_AGE,
    operator: Operator.GREATER_THAN,
    value: 18
  };

  // TypeORM adapter usage
  const typeormBuilder = new ExampleTypeORMConditionAdapterBuilder();
  const typeormAdapter = new TypeORMContextAdapter<User, typeof TestPropertyRefImpl>(
    typeormBuilder, 
    TestPropertyRefImpl
  );

  typeormAdapter.addCondition("nameFilter", nameFilter);
  typeormAdapter.addCondition("ageFilter", ageFilter);

  // Prisma adapter usage
  const prismaBuilder = new ExamplePrismaConditionAdapterBuilder();
  const prismaAdapter = new PrismaContextAdapter<User, typeof TestPropertyRefImpl>(
    prismaBuilder,
    TestPropertyRefImpl
  );

  prismaAdapter.addCondition("nameFilter", nameFilter);
  prismaAdapter.addCondition("ageFilter", ageFilter);

  console.log("✅ TypeScript adapters validation successful!");
  
  return { typeormAdapter, prismaAdapter };
}

export { demonstrateAdapters, ExampleTypeORMConditionAdapterBuilder, ExamplePrismaConditionAdapterBuilder };

